Hardens GB+Tree testing

tinwelint opened the pull request:
By using adversarial page cache and an added access-checking page cache.

Also contains fixes for exposed issues for the added testing.

tinwelint committed:
Access-checking page cache testing utility

for asserting that read cursprs are never used without shouldRetry loops.

tinwelint committed:
Fixing issue with shouldRetry-less reading of tree state pages

reading tree state pages means reading two pages, where each is now read
with its own shouldRetry loop.

tinwelint committed:
GB+Tree tests runs with access checking page cache

ConsistencyChecker was changed to do proper shouldRetry-access as part of this.

tinwelint committed:
TreePrinter do correct shouldRetry access

tinwelint committed:
Better format in print tree

tinwelint committed:
Fixes some javadoc errors

tinwelint committed:
More configurable PageCacheRule

and more unified configurability. Both constructor of the PageCacheRule
as well as the getPageCache method call accepts a PageCacheConfig which
is a builder which can specify things like tracer, page size as well as
whether or not to decorate the page cache in some adversarial wrapper.

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/StandalonePageCacheFactory.java:
```diff
@@ -34,22 +34,27 @@
     }
 
-    public static PageCache createPageCache( FileSystemAbstraction fileSystem, PageCacheTracer tracer )
+    public static PageCache createPageCache( FileSystemAbstraction fileSystem, Integer pageSize,
```
Not too happy about this public method that does null logic on an Integer argument without any comment, I would like one of these to be fixed; for example add a comment, or add more constructors so that you can hide your null-magic in this class.

chrisvest replied:
I agree with this sentiment. You could add another static constructor that takes a `PageSwapperFactory` and a page size, and then move the construction of the default page swapper factory up one level in the stack. The creation & configuration of this page swapper factory can also be extracted into a static helper method. That way you would not need any null-logic at all.

tinwelint replied:
Very much agree

ragadeeshu reviewed (changes requested):
Looks good overall. I have a comment about the changes made to `StandalonePageCacheFactory`. It would also be great if @chrisvest could take a look at `4135a84416c1ab0b0ae0f8b413ef1d6a80cc0248` and how it fits into the page cache surface.

burqen commented:
@ragadeeshu Something like this?

tinwelint commented:
Tools compilation error seems to be unrelated and matches recent changes around configuration refactorings in neo4j

spacecowboy commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/CommonAbstractStoreTest.java:
```diff
@@ -193,7 +193,8 @@ public void recordCursorPinsEachPageItReads() throws Exception
         RecordingPageCacheTracer tracer = new RecordingPageCacheTracer( Pin.class );
-        PageCache pageCache = pageCacheRule.getPageCache( fileSystemRule.get(), tracer, Config.empty() );
+        PageCache pageCache = pageCacheRule.getPageCache( fileSystemRule.get(),
+                config().withTracer( tracer ), Config.empty() );
```
It is true that the Tools compile failure is unrelated but you still have some merge issues. `Config.empty()` no longer exists and the PR builder won't tell you that because it is building your branch and not the merge result. So a rebase/merge is still required.

burqen replied:
Nice catch!

tinwelint replied:
Although it looks like [Config.empty()](https://github.com/neo4j/neo4j/blob/3.2/community/kernel/src/main/java/org/neo4j/kernel/configuration/Config.java#L87) still exists, and this rebased branch compiles and is generally happy now. Thanks for the heads up

tinwelint committed:
GB+Tree tests run with adversarial page cache

also included are fixes for issues exposed by this added testing:

- Removed Layout#read/writeMetaData because there's no need for it
  and it was implemented in the wrong way w/ regards to shouldRetry
- KeySearch#search doesn't throw exception on unexpected end result,
  instead uses one of the high bits to signal success/failure.
- GSPP#pointerState no longer throws exception on unexpected generation
  because it was done inside shouldRetry loop. Now instead reports
  BROKEN state.
- SeekCursor asserts shouldRetry is true after unexpected read,
  otherwise throws exception about tree inconsistency.
- Fixes a bug in AdversarialWritePageCursor#copyTo where this wasn't
  allowed previously, due to an instanceof check.

tinwelint committed:
Moves SeekCursor fullRead check

into branch so that it's only checked if the header wasn't sanely read.
